src/sentry/ratelimits/leaky_bucket.py

```diff
@@ -0,0 +1,212 @@
+from __future__ import annotations
+
+import functools
+import logging
+from collections.abc import Callable, Generator
+from contextlib import contextmanager
+from dataclasses import dataclass
+from time import time
+from typing import Any, ContextManager
+
+from django.conf import settings
+
+from sentry.exceptions import InvalidConfiguration
+from sentry.utils import redis
+
+logger = logging.getLogger(__name__)
+
+
+leaky_bucket_info = redis.load_redis_script("ratelimits/leaky_bucket.lua")
+
+
+@dataclass
+class LeakyBucketLimitInfo:
+    burst_limit: int  # maximum number of requests allowed in a burst
+    drip_rate: int  # number of requests allowed per second
+    last_drip: float = 0  # unix timestamp of the last drip
+    current_level: float = 0  # current level of the bucket
+    wait_time: float = 0  # seconds to wait until next request is allowed
+
+
+class LeakyBucketRateLimiter:
+    NAMESPACE = "leaky_bucket_limiter"
+
+    class LimitExceeded(Exception):
+        def __init__(self, info: LeakyBucketLimitInfo) -> None:
+            self.info = info
+
+    def __init__(self, burst_limit: int, drip_rate: int) -> None:
+        cluster_key = settings.SENTRY_RATE_LIMIT_REDIS_CLUSTER
+        self.client = redis.redis_clusters.get(cluster_key)
+        self.burst_limit = burst_limit
+        self.drip_rate = drip_rate
+
+    def redis_key(self, key: str) -> str:
+        return f"{self.NAMESPACE}:{key}"
+
+    def validate(self) -> None:
+        try:
+            self.client.ping()
+            self.client.connection_pool.disconnect()
+        except Exception as e:
+            raise InvalidConfiguration(str(e))
+
+    def use_and_get_info(self, key: str, timestamp: float | None = None) -> LeakyBucketLimitInfo:
```

Comment:
nit: Could `key` be part of the `__init__` method? That way you don't need to pass the key when calling the various methods.

Reply:
made it optional, also made decorator use original function's fully qualified name as default key